### PR TITLE
Snowflake dialect: Add support for DATABASE ROLE in GRANT/REVOKE

### LINKED ISSUE #5489: Snowflake dialect: GRANT DATABASE ROLE not supported
### Search before asking

- [X] I searched the [issues](https://github.com/sqlfluff/sqlfluff/issues) and found no similar issues.

### What Happened

I have a file containing the following SQL:
```
grant database role dbname.rolename to role public;
```
I ran `sqlfluff parse --dialect snowflake test.sql` and got the following error:
```
==== parsing violations ====
L:   1 | P:   1 |  PRS | Line 1, Position 1: Found unparsable section: 'grant database role
                       | dbname.rolename to r...'
```

### Expected Behaviour

SQLFluff should consider this valid SQL in the Snowflake dialect. Snowflake has documented this syntax here: https://docs.snowflake.com/en/sql-reference/sql/grant-database-role.

### Observed Behaviour

SQLFluff produced a parsing violation for the valid `GRANT DATABASE ROLE` syntax. 

### How to reproduce

Create a file `test.sql` with the following contents:
```
grant database role dbname.rolename to role public;
```
Then, run 
```
sqlfluff parse --dialect snowflake test.sql
```

### Dialect

Snowflake

### Version

```
sqlfluff --version
sqlfluff, version 2.3.5

python --version
Python 3.8.5
```

### Configuration

Default configuration, no `.sqlfluff*` files.

### Are you willing to work on and submit a PR to address the issue?

- [X] Yes I am willing to submit a PR!

### Code of Conduct

- [X] I agree to follow this project's [Code of Conduct](https://github.com/sqlfluff/sqlfluff/blob/main/CODE_OF_CONDUCT.md)

### PR DESCRIPTION
**Pull request by sfc-gh-dgupta**
### Brief summary of the change made
fixes #5489, and related issues with REVOKE and GRANT <permission> TO DATABASE ROLE which weren't discovered at the time I reported the issue. 

### Are there any other side effects of this change that we should be aware of?

Modified existing test fixtures for Snowflake (`grant_revoke.sql`, `grant_revoke.yml` and `create_database_role.yml`).

### Pull Request checklist
- [x] Please confirm you have completed any of the necessary steps below.

- Included test cases to demonstrate any code changes, which may be one or more of the following:
  - `.yml` rule test cases in `test/fixtures/rules/std_rule_cases`.
  - `.sql`/`.yml` parser test cases in `test/fixtures/dialects` (note YML files can be auto generated with `tox -e generate-fixture-yml`).
  - Full autofix test cases in `test/fixtures/linter/autofix`.
  - Other.
- Added appropriate documentation for the change.
- Created GitHub issues for any relevant followup/future enhancements if appropriate.

### CODE REVIEW

File: src/sqlfluff/dialects/dialect_snowflake.py [hunk 1]
```diff
@@ -6067,7 +6091,7 @@ class CreateDatabaseRoleStatementSegment(BaseSegment):
             "IfNotExistsGrammar",
             optional=True,
         ),
-        Ref("ObjectReferenceSegment"),
+        Ref("DatabaseRoleReferenceSegment"),
```
Comment (sfc-gh-dgupta) [inline src/sqlfluff/dialects/dialect_snowflake.py:6094]: This was a small improvement on my previous PR https://github.com/sqlfluff/sqlfluff/pull/5475. If SQLFluff very strictly follows the 1 PR = 1 change rule, I can back this change out.

File: test/fixtures/dialects/snowflake/create_database_role.yml [hunk 2]
```diff
@@ -3,14 +3,14 @@
...
-_hash: d06d6d65331fc37545e9d69c7311a2b0babf375128af2b71e3fb99e547f6f071
+_hash: 8a192cf4ddea53c091d92c24c2029bf1f3807a87327ab497edf6edaf66c2416f
 file:
   statement:
     create_database_role_statement:
     - keyword: create
     - keyword: database
     - keyword: role
-    - object_reference:
+    - database_role_reference:
```
Comment (sfc-gh-dgupta) [inline test/fixtures/dialects/snowflake/create_database_role.yml:13]: See https://github.com/sqlfluff/sqlfluff/pull/5490/files#r1435313776

**Comment by sfc-gh-dgupta** [comment]
If this PR followed Snowflake's documentation as strictly as possible, it would likely introduce a few new statement segments:
1. https://docs.snowflake.com/en/sql-reference/sql/grant-database-role
2. https://docs.snowflake.com/en/sql-reference/sql/grant-database-role-share

in addition to the modifications to the existing `AccessStatementSegment` for grants and revokes. I opted against this approach because it seems the precedent is to keep all GRANT statements, even those with their own doc page like https://docs.snowflake.com/en/sql-reference/sql/grant-role, in the `AccessStatementSegment` class. I'm happy to tweak the approach if my understanding is not right.

File: src/sqlfluff/dialects/dialect_snowflake.py (hunk 1 above)
Comment (greg-finley) [reply src/sqlfluff/dialects/dialect_snowflake.py:6094]: Nope, we can group together related changes where it makes sense. 👍

File: src/sqlfluff/dialects/dialect_snowflake.py [hunk 3]
```diff
@@ -2479,6 +2495,7 @@ class AccessStatementSegment(BaseSegment):
                     "IN",
                     OneOf("DATABASE", "SCHEMA"),
                 ),
+                Sequence("DATABASE", "ROLE"),
```
Comment (greg-finley) [inline src/sqlfluff/dialects/dialect_snowflake.py:2498]: I almost want to make this `DatabaseRoleGrammar` since it's used so much

**Review by greg-finley — APPROVED** [review]
Thanks!

**Pull request merged**